### PR TITLE
Opt

### PR DESCRIPTION
**Pull request by shabanian**
#817

### CODE REVIEW

**Comment by shabanian** [comment]
@adri-romsor

**Comment by nouiz** [comment]
travis is failing:

```
======================================================================

FAIL: theano.tensor.tests.test_opt.test_local_expm1

----------------------------------------------------------------------

Traceback (most recent call last):

  File "/home/travis/miniconda2/envs/pyenv/lib/python2.6/site-packages/nose/case.py", line 197, in runTest

    self.test(*self.arg)

  File "/home/travis/build/Theano/Theano/theano/tensor/tests/test_opt.py", line 6143, in test_local_expm1

    assert numpy.all(f_val == f_test(x_val))

AssertionError
```

It fail when mode=FAST_COMPILE. You can manually include the opt in the test to make sure it get applied.

File: theano/tensor/tests/test_opt.py [hunk 1]
```diff
@@ -6122,6 +6123,38 @@ def test3(self):
...
+    t = T.exp(x) - x
+    s = T.exp(u) - numpy.ones((4, 3)).astype(config.floatX)
+
+    f = function([x], y)
+    g = function([x], z)
+    h = function([x], t)
+    r = function([u], s)
+    x_val = numpy.random.rand(4, 3).astype(config.floatX)
+    f_val = f(x_val)
+    f_test = function([x], T.expm1(x))
```
Comment (nouiz) [inline theano/tensor/tests/test_opt.py]: pass mode=theano.compile.get_default_mode().including('local_expm1')

It should fix travis.

File: theano/tensor/opt.py [hunk 2]
```diff
@@ -3576,10 +3576,37 @@ def local_join_make_vector(node):
...
+    """
+    This optimization detects exp(a)-1 and converts this to expm1(a).
+    """
+    if (isinstance(node.op, T.Elemwise) and
+            isinstance(node.op.scalar_op, theano.scalar.basic.Sub)):
+        in1, in2 = node.inputs
+        out = node.outputs[0]
+
+        if (in1.owner and isinstance(in1.owner.op, T.Elemwise) and isinstance(in1.owner.op.scalar_op, theano.scalar.basic.Exp) and
+                T.extract_constant(in2) == 1):
```
Comment (nouiz) [inline theano/tensor/opt.py]: Add the parameter only_process_constants=True to extract_constant().

**Comment by shabanian** [comment]
We set it to False so that it would dig through the dimshuffles, in case we have an array of ones (either as part of the initial call or as a result of a broadcastable dimension). Is that correct or are we missing something?

**Pull request merged**